### DISCUSSION TITLE
How to obtain the first order wave excitation force with different degrees of freedom in different directions, such as (sway, surge, pitch)

### DISCUSSION THREAD

**Discussion by icnosvbovwb** [OpenFAST/openfast#1472] [Q&A]
Dear @jjonkman 
Q1:How to obtain the first order wave excitation force with different degrees of freedom in different directions, such as (sway, surge, pitch)?    The horizontal ordinate is the frequency.  The longitudinal coordinates is the Surge Force.
Just like this picture
![image](https://user-images.githubusercontent.com/107483730/221891835-8b0ee512-1841-4faf-bea6-aceeeefc5d82.png)
Q2:How to obtain RAO with different degrees of freedom in different directions, such as (sway, surge, pitch)?  The horizontal ordinate is the frequency.  The longitudinal coordinates is the Surge Rao.
![image](https://user-images.githubusercontent.com/107483730/221892842-85a14e25-9bc6-4430-b2ae-91708ae71e2e.png)

**Comment by jjonkman** [comment]
Dear @icnosvbovwb,

It sounds like you are asking about how to solve standard hydrodynamic wave-body radiation and diffraction problems in the frequency domain.  You can use commercial software such as WAMIT or AQWA or open-source software such as NemoH, Capytaine, or HAMS for that.  We often use WAMIT or HAMS at NREL.

FYI: The frequency-dependent radiation and diffraction loads are inputs to the HydroDyn module of OpenFAST when using the potential-flow solution; within HydroDyn, these are transformed to the time domain via frequency-to-time domain transforms.

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
I have downloaded HAMS and run it according to Manual to get the results. Please tell me where to set the direction.
And what does the title of these columns of data mean？
![image](https://user-images.githubusercontent.com/107483730/223115890-af2fda65-5b7b-404c-a5f3-0c3fe0549b11.png)

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

Your data seems to be from a WAMIT-formatted wave-excitation (_*.3_) file, where the second column contains the wave direction.  The format of this file is documented in the online WAMIT documentation: https://www.wamit.com/manual6.4/Chap4.pdf.

Best regards,

**Reply by icnosvbovwb** [reply]
Dear @jjonkman 
Q1:I want to know what the first column represents. I guess it's frequency. Its range is 0.5-1.5. How to set 0rad/s-3rad/s ?
Q2:The direction interval of the second column is 10 degrees. How can I set the interval to 5 degrees? 
Q3:What does 1 2 3 4 5 6 in the third column mean? I guess it is six degrees of freedom (surge, sway, heave, roll, pitch, yaw)
Q4:What does the fourth to seventh columns mean?

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

The WAMIT documentation I linked to explains what the columns are:

`OPTN.3: PER BETA I Mod(Xi) Pha(Xi) Re(Xi) Im(Xi)`

The first column is period, the second column is direction, the third column is the load component, and columns 4-7 are the amplitude, phase, real, and imaginary components of the normalized wave-excitation load.

I have not used HAMS myself, so, can't comment on how to define these values in HAMS.  I would expect the HAMS documentation to explain that or perhaps someone else can comment.

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
![image](https://user-images.githubusercontent.com/107483730/223650459-bae9e02c-9b2c-4982-ba2e-8c13beeb4a1f.png)
![image](https://user-images.githubusercontent.com/107483730/223650261-653e81ee-d6f3-469c-945b-c02887ade70f.png)
If the right side of the first figure 1-6 represents the amplitude, phase, real part and imaginary part respectively, which column should I take in the red part of the second figure?

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

You don't say what you are plotting in your figure, but presumably you are plotting the amplitude of the surge force (column 4 of the _.3_ file associated with load component 1).

FYI: It looks like your figure gives the surge force in Newtons, whereas the WAMIT output is nondimensional.  The nondimensionalization of the WAMIT output is explained in the WAMIT documentation I linked to.

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
![image](https://user-images.githubusercontent.com/107483730/223729592-888a9cd6-b148-4189-98c5-6f9397141af4.png)
Yes, Your understanding is right. But how to transform nondimensional  to Newtons unit?
And another question is that I will set the Number_of_frequencies in the HAMS control file is 35, that is, 0rad/s-3rad/s, with an interval of 0.1 in the middle, so the left circle should be the frequency domain, not the period, right?

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

See the WAMIT documentation I linked to.  You can redimensionalize using the water density, gravity, wave amplitude, and length scale.

Best regards,

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
![image](https://user-images.githubusercontent.com/107483730/223744271-c9b53dbd-97d1-4614-8354-6abc1bf1c240.png)
Q1:This is a file in HAMS output format. What does OAMASS1 in the first picture mean, and what does the 8 columns of data in it mean respectively?
![image](https://user-images.githubusercontent.com/107483730/223744116-8bc97abc-b566-47aa-80bf-b969b389a4c3.png)
Q2:What does ODAMPING1 in the second picture mean, and what does the 4 columns of data in it mean respectively?

**Reply by jjonkman** [reply]
Sorry,but I'm not familiar with these files.  Perhaps someone else can comment.

**Reply by YingyiLiu** [reply]
Dear @icnosvbovwb,

The screenshot you posted here is from the HAMS-format output. I usually use these simple-format results to do time-domain simulations coded by myself. Since you want to calculate motion RAOs, I suggest you use the Wamit-format output file *.4 generated by HAMS. An example here: https://github.com/YingyiLiu/HAMS/tree/master/CertTest/DeepCwind/Output/Wamit_format. The format is then exactly as that described in Wamit v7.4 user manual Page 88:     OPTN.4: PER BETA I Mod(Ksi_i) Pha(Ksi_i) Re(Ksi_i) Im(Ksi_i).

**Reply by YingyiLiu** [reply]
Dear @icnosvbovwb,

I can tell you that OAMASSn, ODAMPINGn and OEXFORn (n=1,2...,6) mean the added mass, the radiation damping and the complex excitation force in the nth DoF of the floater, respectively. The 4 columns of the data in the 2nd picture are wave number, wave angular frequency, real part of the wave force, and imaginary part of the wave force.

**Reply by icnosvbovwb** [reply]
Dear @YingyiLiu 
![image](https://user-images.githubusercontent.com/107483730/224485003-53ba9816-6b09-42d0-bc37-6260646198e4.png)
My purpose is to get the data shown in the picture, but I'm not sure whether to use. 3 file or. 4 file now. They are all in Wamit_ Format file. If one of the files is used, the fourth column of data is dimensionless. I found the conversion formula, in which we need to determine the density, gravity acceleration, and amplitude A. Now it is amplitude A. I don't know how to calculate it.

**Comment by icnosvbovwb** [comment]
Dear @YingyiLiu
Can you tell me what is the mean in 8 columns in OAMASS1 and 4 columns in ODAMPING1?

**Reply by YingyiLiu** [reply]
OAMASSn and ODAMPINGn both have 8 columns of data. Take OAMASS1 for example. The 8 columns refer to wave number, wave angular frequency, A11, A12, A13, A14, A15, A16, respectively. Here Aij means the added mass of the ith DoF of the floater due to its jth-DoF motion.

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
![image](https://user-images.githubusercontent.com/107483730/224309734-612a0fa2-f301-4731-875a-085077c9eed2.png)
![image](https://user-images.githubusercontent.com/107483730/224309814-c2659516-0714-4169-951c-88824dde444b.png)
Can you tell me  the incident-wave amplitude A in DEFINITION OF QUANTITIES EVALUATED BY WAMIT is Significant wave height of incident waves in HydroDyn Input File? Because I transform nondimensional to Newtons unit using ρ=1025,g=9.806,A=??,Lm=1.

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

The WAMIT output is in the frequency domain and nondimensional.  At a given wave frequency, the wave amplitude is equivalent to half the wave height of a regular wave at that frequency.  Linear wave-excitation loads will be proportional to this wave amplitude.

Best regards,

**Reply by zanderder** [reply]
Dear @jjonkman 

Regarding to your comment 'Linear wave-excitation loads will be proportional to this wave amplitude', I agree with you, but I have a question. By converting to dimensional excitation force, the nondimensional value has been multiplied with 'rho * g * A * L^2'. Is it necessary to multiply an additional 'A' to get the excitation force? F_ex=X_i * A * cos(omega*t+phi)=rho * g * L^2 * Mod(X_i_non) * cos(omega * t+phi)? where X_i_non is the nondimensional excitation force amplitude or the fourth column of *.3 output file. Thank you.

Best regards

**Reply by jjonkman** [reply]
Dear @zanderder,

I'm not sure I fully understand your question, but there is only one multiplication by the wave-elevation "A" when calculating the wave-excitation loads based on the normalized wave-excitation data stored in WAMIT format.

Best regards,

**Reply by zanderder** [reply]
Dear @jjonkman,

Great to hear back from you so quick! Maybe I did make it clear. In WAMIT manual, the redimensionalization is done through the following formula:

![image](https://github.com/user-attachments/assets/ac241e11-6a6a-4367-b728-c27f56ac8b05)

So, after reading your comments above, I was wondering whether X_i should be multiplied by an additional "A" for scaling. It seems that by using the above fomula, the excitation force has already be both dimensionalized and scaled. Is it correct?

Thank you.

Best regards,

**Reply by zanderder** [reply]
"rho * g * A * L^2" = kg/m^3 * m/s^2 * m * m^2 = kg m/s^2 = N 
This makes sense

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
![image](https://user-images.githubusercontent.com/107483730/224332289-3821f22c-8cd1-4c9c-b94d-69006b81b63d.png)
I don't quite understand what you said. I got Wamit_ Format file through HAMS, and the fourth column of. 3 file is converted to Newton units, so now only the amplitude A is not sure where to find it?

**Reply by jjonkman** [reply]
Dear @icnosvbovwb,

Again, I'm not familiar with HAMS, but if the HAMS output is dimensional, the wave amplitude must be specified within HAMS.  Or perhaps the output is not in Newton, but in Newton per meter (i.e., force per unit wave amplitude).

Best regards,

**Reply by YingyiLiu** [reply]
Just a supplement. As I mentioned above, HAMS can generate results in three formats. The HAMS-format outputs are dimensional which I usually use to do some straightforward time-domain simulations. For OpenFAST users, I suggest they use the Wamit-format output files if they are running HAMS, since the Wamit format is standard and well-documented.

**Comment by icnosvbovwb** [comment]
Dear @jjonkman 
Thank you for your reply. I want to know which of the software you recommend are most familiar to you, and the functions I want are ｔｈｅｓｅ：First-order wave excitation force, added mass, radiation damping, etc. corresponding to different incident wave angle, different water depth, different frequency domain, etc

**Reply by jjonkman** [reply]
I'm personally familiar with WAMIT, but I'm sure there are other experts that use the OpenFAST repository that know other software.

**Comment by icnosvbovwb** [comment]
Dear @jjonkman
Can you give me a link with WAMIT　to download？Because I can't find where to download。

**Reply by jjonkman** [reply]
I've already included a link to the WAMIT website above, but WAMIT is not a free software.

**Comment by icnosvbovwb** [comment]
Dear @YingyiLiu 
![image](https://user-images.githubusercontent.com/107483730/224486000-34e26e34-924a-4c0a-9b63-94ffb29a8523.png)
Another question is how to use the red range of the mark. Can you give the difference between them?
Because I set the ControlFile of HAMS according to the first cycle of the. 3 file in the HydroData folder of OPENFAST ( Input_ frequency_ type=Output_ frequency_ Type=4 ) . But the result is different. The results are shown in the figure:
![image](https://user-images.githubusercontent.com/107483730/224548587-9d8601cf-d816-4359-b6ef-8f42e18603c0.png)
![image](https://user-images.githubusercontent.com/107483730/224548756-ac33e299-2f42-4039-890d-cf66ac091185.png)

**Reply by YingyiLiu** [reply]
A negative value of water depth, like -200m, means deep water. That is perhaps the reason for the difference. You will need to set the water depth as 200 m and run again the computation.

**Comment by icnosvbovwb** [comment]
Dear @YingyiLiu 
How does HMAS calculate the amplitude corresponding to each angular frequency? When I know the amplitude, I can know the force.

**Reply by YingyiLiu** [reply]
In a frequency-domain preprocessor, such as Wamit, Hams, etc., the amplitude is normally set as 1.0m to calculate added masses, radiation dampings and wave excitation forces. These are all linear transfer functions.

**Comment by icnosvbovwb** [comment]
Dear @YingyiLiu
![image](https://user-images.githubusercontent.com/107483730/225306793-3fae0101-3dea-48ba-9a46-2449d4dd7513.png)
I draw the wave incidence direction of the wind turbine in this way. Is this angle correct?

**Reply by YingyiLiu** [reply]
Not correct. The angle you notated as -180 deg should be 0 deg; the angle you notated as 0 deg should be 180 deg; the angle you notated as 90 deg should be 270 deg, and so on ...

**Reply by icnosvbovwb** [reply]
Dear @YingyiLiu 
![image](https://user-images.githubusercontent.com/107483730/226574996-7bd455fd-5ab6-4da0-99d8-c71d7d056ff2.png)
Can other expressions be expressed like this pitcure? For example, Can it express 90 deg as - 90 deg/270deg? There are also 180 degrees and - 180 degrees on the same axis. Am I correct in expressing this?

**Comment by icnosvbovwb** [comment]
Dear @YingyiLiu 
RAO
![image](https://user-images.githubusercontent.com/107483730/226539766-1c478d43-66af-4496-95b2-37d6624a9406.png)
First order wave excitation force
![image](https://user-images.githubusercontent.com/107483730/226541189-11f7ae85-d73f-4b30-b4cc-4e755077e13e.png)
WaveDamping
![image](https://user-images.githubusercontent.com/107483730/226541278-d22e1a01-b812-4f35-aef1-a0e41c56db63.png)
AddedMass
![image](https://user-images.githubusercontent.com/107483730/226541321-83d6e50b-5cf0-4624-be34-b4aa7fb2084c.png)

This is first order wave excitation force and RAO with different degrees of freedom in different directions of OC4. I observed that the RAO of the pitch is not normal. Can you find out why? And overall, is it correct?

**Reply by YingyiLiu** [reply]
Thanks for the comment. Could you please indicate more details about what you think is not normal in the pitch RAO? I don't find something abnormal actually.

**Reply by icnosvbovwb** [reply]
Dear @YingyiLiu 
![image](https://user-images.githubusercontent.com/107483730/226575357-c4387e7b-3ec6-4f2d-bb69-8268cfc91e71.png)
![image](https://user-images.githubusercontent.com/107483730/226575437-335e23c6-5712-4981-b5f3-7c85e1a637ac.png)
As you can see, my results are not quite the same as those of others, and I am not sure if they are correct? Because I calculated the pitch frequency for free attenuation to be 0.038Hz, which is 0.23864rad/s. According to theory, the peak value should be around 0.23, but my results show that there are two peaks, one is 0.06rad/s, and the other is more than 0.5rad/s, and the value is very small, only 0.012deg/m.

**Reply by YingyiLiu** [reply]
It looks strange to me that the two results are so different. As reported by many others, HAMS can generate almost the same result as WAMIT. Motion RAOs are solutions to the motion equation which involves mass matrix, added mass matrix, radiation damping matrix, hydrostatic restoring matrix, and wave excitation force matrix. I suggest to you compare each matrix one by one and then you will find out where the problem occurs.

**Comment by icnosvbovwb** [comment]
Dear @YingyiLiu 
![image](https://user-images.githubusercontent.com/107483730/226597539-5f133212-3a17-4479-ab7f-0a2b96c26c7e.png)
![image](https://user-images.githubusercontent.com/107483730/226598131-eb137f84-61b2-41e5-92dd-6fc9492babfe.png)
![image](https://user-images.githubusercontent.com/107483730/226602113-850199a6-0af2-4288-9be8-ebd46e07a239.png)

![image](https://user-images.githubusercontent.com/107483730/226599643-180749a5-bddc-4ee6-baf4-36c618b275be.png)
![image](https://user-images.githubusercontent.com/107483730/226599824-e0bb4b47-49c4-4838-ba0e-2fbe9e6afb2b.png)
![image](https://user-images.githubusercontent.com/107483730/226599935-9444c80b-5155-40f0-ad0f-fc95e5b64b7f.png)

The spar/TLP/Barge Buoy. 3 result I calculated is also different from that of OPENFAST spar/TLP/Barge. 3. I am not sure where need to set it? Or is there any difference between the HAMS model and OPENFATS medel?

**Reply by YingyiLiu** [reply]
Sorry, it's not possible for me to find out the answer with the limited information given above. Please check each matrix one by one as I mentioned above.

**Reply by icnosvbovwb** [reply]
Dear @YingyiLiu 
![image](https://user-images.githubusercontent.com/107483730/227705632-1be7d13a-693e-4a60-9b72-afebc6441f09.png)
![image](https://user-images.githubusercontent.com/107483730/227705650-cf6aca0e-3004-4172-a84d-9b0009275abd.png)
![image](https://user-images.githubusercontent.com/107483730/227705666-602f4fd0-8339-4bc3-92f3-e25d7defd686.png)
I am using the downloaded Hydrostatic.IN file directly, and I have not made any changes. I only make changes to the cycle and wave direction in ControlFile.in.

**Reply by icnosvbovwb** [reply]
Dear @YingyiLiu 
Can you provide the latest versions of four models (Hydrostatic.in, HullMesh.PNL, and WaterplaneMesh.PNL). I guess there should be some problems with the modeling of the four models, which can lead to inconsistent results with the. 3 file in openfast
